### ISSUE TITLE
IceNet2 model development codebase updates

### ISSUE THREAD

**Issue by tom-andersson** [alan-turing-institute/IceNet-Project#23]
Roughly in order of importance with a bold section for each codebase update!

**Linear trend class and preprocessing**:
* New LinearTrendModel() class in models.py (has logic for missing SIC dates from either hemisphere): https://github.com/tom-andersson/icenet2/blob/9d3e6dd77461b9e28d2dca799ba656b877394d99/icenet2/models.py#L160
* Predict script allows for linear trend benchmark: https://github.com/tom-andersson/icenet2/commit/3132ec7f7e0f72105db56649b031f78996425983#diff-8613ab00b1c68e2133f3e3a069b134e02c0d4481b958146bd14ec4905088fea6
* Analyse script allows for linear trend benchmark: https://github.com/tom-andersson/icenet2/commit/d8b4d10591f2b1ee6652dc6346723b39f4d4ce7a#diff-f9256a5444b919139b4083471cfd33cf7712d3416fc17f007284dacd56f0214c

**IceNet2DataPreprocessor updates**: https://github.com/tom-andersson/icenet2/commit/4204a2ab0444af8275914ab1d0f0178e7ade2639#diff-48347981dc5c37725cd87f5dc124fefab2f515c9dd42b567234819410ddd6f56
* Now just uses start/end dates of forecast initialisations to specify train/val/test (rather than passing a list of timestamps) for simplicity
* Now uses cached normalisation parameters and climatologies (doesn't recompute if the parameters/climatology already exists)
* Logic for building linear trend forecast da for preprocessing: https://github.com/tom-andersson/icenet2/blob/9d3e6dd77461b9e28d2dca799ba656b877394d99/icenet2/utils.py#L347

**IceNet2DataLoader updates**
* data_generation now outputs third tensor for `sample_weight` for implicit grid-cell-wise weighting during training (rather than hacky solution of including as a separate channel of `y` and using custom metrics): https://github.com/tom-andersson/icenet2/commit/6344487805546df2a7559352e9d075eaa5610499#diff-17ed9413e234cebb66683e18ebf3d08335ef0ab7df06c9cbcdac2757440fe54c
* New `refresh` method allows dataloader config to be updated by manually passing in a dictionaryof configuration (currently just used for varying the `max_lag` hyperparameter using a single base dataloader config): https://github.com/tom-andersson/icenet2/blob/9d3e6dd77461b9e28d2dca799ba656b877394d99/icenet2/utils.py#L628

**Testing out subclassing tf.metrics and tf.losses in place of custom metrics (for MAE, RMSE, MSE)**: https://github.com/tom-andersson/icenet2/commit/12c5bee17f0a758d33c24f6220c4fc62d3873bae#diff-48347981dc5c37725cd87f5dc124fefab2f515c9dd42b567234819410ddd6f56
* Now uses implicit sample_weights weighting from new third element of `(X, y, sample_weight)` in `Model.fit`and `Model.evaluate` to weight metrics

**New sweep.yaml for hyperparameter tuning with wandb***: https://github.com/tom-andersson/icenet2/blob/da705d9d1193c10fafdbb58c56f443164d746e35/icenet2/sweep.yaml

**Train script updates**
* Trying to enable XLA for faster training (although doesn't seem to have made a difference yet): https://github.com/tom-andersson/icenet2/blob/b768256b1e7ba1dc7900e10bd9fff6839168bef2/icenet2/train_icenet2.py#L20
* Option to train on pre-generated TFRecords dataset 

**Script to generate TFRecords training dataset**: https://github.com/tom-andersson/icenet2/blob/main/icenet2/gen_tfrecords_dataset.py
* Optionally overriding `max_lag` hyperparameter in dataloader config (using command line entry) and saving dataset to `<tfrecords_folder>/<max_lag>`, to be used as wandb hyperparameter tuning

**Script to download ocean reanalysis data (ORAS5)**: https://github.com/tom-andersson/icenet2/blob/main/icenet2/download_and_regrid_daily_oras5_data.py
* With bash script to kick off downloads for multiple hemispheres/variables/years in parallel without exceeding download limit: https://github.com/tom-andersson/icenet2/blob/main/download_oras5_data_in_parallel.sh

**Script to rotate arbitrary vector data (e.g. wind or ocean velocity) from lat-lon to EASE**: https://github.com/tom-andersson/icenet2/blob/main/icenet2/rotate_vector_data.py

**SLURM script for wandb hyperparameter tuning on BAS GPU node**: https://github.com/tom-andersson/icenet2/blob/main/slurm_sweep.sh

**Environments**:
* Locked env: https://github.com/tom-andersson/icenet2/blob/main/environment.locked.yml
* Trying to set up TF2.4 environment to see if it leads to training speedup. Direct dependency file: https://github.com/tom-andersson/icenet2/blob/main/environment_tf24.yml

**from datetime import datetime replace with pd.Timestamp throughout codebase**
* Custom `filled_daily_dates` replaced with `pd.date_range` throughout

**preproc script: new ocean variable and siconca linear trend*: https://github.com/tom-andersson/icenet2/commit/31abd75ae45e67cf0a95280a8c1f9bee6310239e#diff-48347981dc5c37725cd87f5dc124fefab2f515c9dd42b567234819410ddd6f56

**Learning rate schedule (exponential decay)**: https://github.com/tom-andersson/icenet2/blob/9d3e6dd77461b9e28d2dca799ba656b877394d99/icenet2/utils.py#L1272
* Off in train_icenet2.py script by default, but can be used by changing `--lr_10e_decay_fac` input hyperparam: https://github.com/tom-andersson/icenet2/blob/b768256b1e7ba1dc7900e10bd9fff6839168bef2/icenet2/train_icenet2.py#L33

**Comment by JimCircadian** [comment]
IceNet2DataLoader updates are done with respect to the sample weightings. In terms of configuration updates, these are natively supported anyway through mechanisms of the library. 

wandb integration for data push is available. I'm not doing sweeps (leaving that to Tom as he develops the model) so no need to progress any further from there.

Script to download ocean reanalysis data (ORAS5) - moving this to a notebook to illustrate how to extend the library, so encompassing in another ticket
